GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/NewProductUpgradeTests.cs

```diff
@@ -0,0 +1,102 @@
+            // without requiring authorization.
+            this.credentialStoreMock = new Mock<ICredentialStore>(MockBehavior.Strict);
+        }
+
+        [TestCase]
+        public void CreatesNuGetUpgraderWhenConfigured()
+        {
+            this.gvfsConfig = this.BuildNuGetGvfsConfig();
+            ProductUpgrader.TryCreateUpgrader(this.tracer, this.fileSystemMock.Object, this.gvfsConfig, this.credentialStoreMock.Object, false, false, out ProductUpgrader productUpgrader, out string error);
+
+            productUpgrader.ShouldNotBeNull();
+            productUpgrader.ShouldBeOfType<NuGetUpgrader>();
+        }
+
+        [TestCase]
+        public void CreatesGitHubWhenConfigured()
+        {
+            this.gvfsConfig = this.BuildGitHubtGvfsConfig();
+            ProductUpgrader.TryCreateUpgrader(this.tracer, this.fileSystemMock.Object, this.gvfsConfig, this.credentialStoreMock.Object, false, false, out ProductUpgrader productUpgrader, out string error);
+
+            productUpgrader.ShouldNotBeNull();
+            productUpgrader.ShouldBeOfType<GitHubUpgrader>();
+        }
+
+        [TestCase]
+        public void NoUpgraderWhenNuGetFeedMissing()
+        {
+            this.gvfsConfig = this.BuildNuGetConfigMissingUpgradeFeed();
+            ProductUpgrader.TryCreateUpgrader(this.tracer, this.fileSystemMock.Object, this.gvfsConfig, this.credentialStoreMock.Object, false, false, out ProductUpgrader productUpgrader, out string error);
+
+            productUpgrader.ShouldBeNull();
+        }
+
+        [TestCase]
+        public void NoUpgraderWhenNuGetPackNameMissing()
+        {
+            this.gvfsConfig = this.BuildNuGetConfigMissingPackageName();
+            ProductUpgrader.TryCreateUpgrader(this.tracer, this.fileSystemMock.Object, this.gvfsConfig, this.credentialStoreMock.Object, false, false, out ProductUpgrader productUpgrader, out string error);
+
+            productUpgrader.ShouldBeNull();
+        }
+
+        private MockLocalGVFSConfig BuildNuGetGvfsConfig()
+        {
+            MockLocalGVFSConfig gvfsConfig = new MockLocalGVFSConfig();
+            gvfsConfig.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, "slow", out _);
+            gvfsConfig.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedPackageName, "package", out _);
+            gvfsConfig.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedUrl, "https://pkgs.dev.azure.com/contoso/", out _);
+            return gvfsConfig;
+        }
+
+        private MockLocalGVFSConfig BuildNuGetConfigMissingUpgradeFeed()
+        {
+            MockLocalGVFSConfig gvfsConfig = new MockLocalGVFSConfig();
+            gvfsConfig.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeRing, "slow", out _);
+            gvfsConfig.TrySetConfig(GVFSConstants.LocalGVFSConfig.UpgradeFeedPackageName, "package", out _);
+            return gvfsConfig;
+        }
+
+        private MockLocalGVFSConfig BuildNuGetConfigMissingPackageName()

```

Comment:
Do we need to test missing `UpgradeRing`?

Reply:
Not at this time - Upgrade ring is not used by `NuGetUpgrader`.  In the main workflow, we expect it to be present, but ignored by the `NuGetUpgrader` logic. `NuGetUpgrader` logic should work, even if it is not present.

Reply:
>`NuGetUpgrader` logic should work, even if it is not present.

Could/should we validate that by having a test with no `UpgradeRing`?